File: CS/11_SecurityAndSignatures/SignInSignatureField.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using Spire.Pdf;
using Spire.Pdf.Widget;
using Spire.Pdf.Security;
using Spire.Pdf.Graphics;

namespace SignInSignatureField
{
    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();
        }

        private void button1_Click(object sender, EventArgs e)
        {
            // Load a PDF document from the disk
            PdfDocument doc = new PdfDocument();
            doc.LoadFromFile("../../../../../../Data/SignatureField.pdf");

            // Retrieve the form widgets from the document
            PdfFormWidget widgets = doc.Form as PdfFormWidget;

            // Iterate through each field widget in the form
            for (int i = 0; i < widgets.FieldsWidget.List.Count; i++)
            {
                // Get the current field widget
                PdfFieldWidget widget = widgets.FieldsWidget.List[i] as PdfFieldWidget;

                // Check if the field widget is a signature field
                if (widget is PdfSignatureFieldWidget)
                {
                    // Get the name of the signature field
                    string name = widget.Name;
                    // Cast the widget to a PdfSignatureFieldWidget
                    PdfSignatureFieldWidget signWidget = widget as PdfSignatureFieldWidget;

                    // Open the Windows certificate store for read-only access
                    System.Security.Cryptography.X509Certificates.X509Store store = new System.Security.Cryptography.X509Certificates.X509Store(System.Security.Cryptography.X509Certificates.StoreLocation.CurrentUser);
                    store.Open(System.Security.Cryptography.X509Certificates.OpenFlags.ReadOnly);

                    // Manually select a certificate from the store
                    System.Security.Cryptography.X509Certificates.X509Certificate2Collection sel = System.Security.Cryptography.X509Certificates.X509Certificate2UI.SelectFromCollection(store.Certificates, null, null, System.Security.Cryptography.X509Certificates.X509SelectionFlag.SingleSelection);

                    // Create a PdfCertificate object using the certificate data from the selected certificate
                    PdfCertificate cert = new PdfCertificate(sel[0].RawData);

                    // Create a PdfSignature object using the document, page, certificate, field name, and signature field widget
                    PdfSignature signature = new PdfSignature(doc, signWidget.Page, cert, name, signWidget);

                    // Load the sign image source
                    signature.SignImageSource = PdfImage.FromFile("../../../../../../Data/E-iceblueLogo.png");

                    // Set the graphics mode for display
                    signature.GraphicsMode = GraphicMode.SignImageAndSignDetail;

                    // Set the label and value for the signer's name
                    signature.NameLabel = "Signer:";
                    signature.Name = sel[0].GetNameInfo(System.Security.Cryptography.X509Certificates.X509NameType.SimpleName, true);

                    // Set the label and value for the contact information
                    signature.ContactInfoLabel = "ContactInfo:";
                    signature.ContactInfo = signature.Certificate.GetNameInfo(System.Security.Cryptography.X509Certificates.X509NameType.SimpleName, true);

                    // Set the label and value for the signing date
                    signature.DateLabel = "Date:";
                    signature.Date = DateTime.Now;

                    // Set the label and value for the signing location
                    signature.LocationInfoLabel = "Location:";
                    signature.LocationInfo = "Chengdu";

                    // Set the label and value for the reason of signing
                    signature.ReasonLabel = "Reason: ";
                    signature.Reason = "The certificate of this document";

                    // Set the label and value for the distinguished name (DN) of the certificate issuer
                    signature.DistinguishedNameLabel = "DN: ";
                    signature.DistinguishedName = signature.Certificate.IssuerName.Name;

                    // Set the document permissions and mark the document as certified
                    signature.DocumentPermissions = PdfCertificationFlags.AllowFormFill | PdfCertificationFlags.ForbidChanges;
                    signature.Certificated = true;

                    // Set the fonts for the sign details and sign name, if not set, default ones will be applied
                    signature.SignDetailsFont = new PdfFont(PdfFontFamily.TimesRoman, 10f);
                    signature.SignNameFont = new PdfFont(PdfFontFamily.Courier, 15);

                    // Set the sign image layout mode to None
                    signature.SignImageLayout = SignImageLayout.None;
                }
            }

            // Save the modified PDF document to a file
            string output = "SignWithSmartCardUsingSignatureField_out.pdf";
            doc.SaveToFile(output);

            //Launch the result file
            PDFDocumentViewer(output);
        }

        private void PDFDocumentViewer(string fileName)
        {
            try
            {
                System.Diagnostics.Process.Start(fileName);
            }
            catch { }
        }
    }
}

File: VB/11_SecurityAndSignatures/SignInSignatureField.vb
Imports Spire.Pdf
Imports Spire.Pdf.Widget
Imports Spire.Pdf.Security
Imports Spire.Pdf.Graphics
Imports System.Security.Cryptography.X509Certificates

Namespace SignInSignatureField
    Partial Public Class Form1
        Inherits Form
        Public Sub New()
            InitializeComponent()
        End Sub

        Private Sub button1_Click(ByVal sender As Object, ByVal e As EventArgs) Handles button1.Click
            ' Create a new PdfDocument object
            Dim doc As New PdfDocument()

            ' Load the PDF document from the specified file path
            doc.LoadFromFile("../../../../../../Data/SignatureField.pdf")

            ' Get the form widget from the document
            Dim widgets As PdfFormWidget = TryCast(doc.Form, PdfFormWidget)

            ' Iterate through each field widget in the form
            For i As Integer = 0 To widgets.FieldsWidget.List.Count - 1
                Dim widget As PdfFieldWidget = TryCast(widgets.FieldsWidget.List(i), PdfFieldWidget)
                If TypeOf widget Is PdfSignatureFieldWidget Then
                    ' Get the field name
                    Dim name As String = widget.Name
                    Dim signWidget As PdfSignatureFieldWidget = TryCast(widget, PdfSignatureFieldWidget)

                    ' Open the windows certificate store in read-only mode
                    Dim store As New X509Store(StoreLocation.CurrentUser)
                    store.Open(OpenFlags.ReadOnly)

                    ' Manually select the certificate from the store
                    Dim sel As X509Certificate2Collection = X509Certificate2UI.SelectFromCollection(store.Certificates, Nothing, Nothing, X509SelectionFlag.SingleSelection)

                    ' Create a certificate using the selected certificate data
                    Dim cert As New PdfCertificate(sel(0).RawData)

                    ' Create a signature with the certificate and the signature field
                    Dim signature As New PdfSignature(doc, signWidget.Page, cert, name, signWidget)

                    ' Load the sign image source
                    signature.SignImageSource = PdfImage.FromFile("../../../../../../Data/E-iceblueLogo.png")

                    ' Set the graphics display mode (default if not set)
                    signature.GraphicsMode = GraphicMode.SignImageAndSignDetail
                    signature.NameLabel = "Signer:"

                    ' Set the signer name and contact information
                    signature.Name = sel(0).GetNameInfo(X509NameType.SimpleName, True)
                    signature.ContactInfoLabel = "ContactInfo:"
                    signature.ContactInfo = signature.Certificate.GetNameInfo(X509NameType.SimpleName, True)

                    ' Set the date of the signature
                    signature.DateLabel = "Date:"
                    signature.Date = Date.Now

                    ' Set the location information
                    signature.LocationInfoLabel = "Location:"
                    signature.LocationInfo = "Chengdu"

                    ' Set the reason for the signature
                    signature.ReasonLabel = "Reason: "
                    signature.Reason = "The certificate of this document"

                    ' Set the distinguished name label and value
                    signature.DistinguishedNameLabel = "DN: "
                    signature.DistinguishedName = signature.Certificate.IssuerName.Name

                    ' Set the document permissions and mark as certified
                    signature.DocumentPermissions = PdfCertificationFlags.AllowFormFill Or PdfCertificationFlags.ForbidChanges
                    signature.Certificated = True

                    ' Set custom fonts for sign details and sign name (default if not set)
                    signature.SignDetailsFont = New PdfFont(PdfFontFamily.TimesRoman, 10.0F)
                    signature.SignNameFont = New PdfFont(PdfFontFamily.Courier, 15)

                    ' Set the sign image layout mode to None
                    signature.SignImageLayout = SignImageLayout.None
                End If
            Next i

            ' Save the modified PDF document to the specified output file path
            Dim output As String = "SignWithSmartCardUsingSignatureField_out.pdf"
            doc.SaveToFile(output)

            ' Close the PDF document
            doc.Close()

            ' Launch the result file
            PDFDocumentViewer(output)
        End Sub

        Private Sub PDFDocumentViewer(ByVal fileName As String)
            Try
                Process.Start(fileName)
            Catch
            End Try
        End Sub
    End Class
End Namespace
